pkg/controller/azvolume.go

```diff
@@ -81,260 +82,303 @@ func (r *ReconcileAzVolume) Reconcile(ctx context.Context, request reconcile.Req
 		return reconcile.Result{Requeue: true}, err
 	}
 
-	if now := metav1.Now(); azVolume.ObjectMeta.DeletionTimestamp.Before(&now) {
-		// azVolume deletion
-		klog.Infof("Beginning deletion of AzVolume object")
-		if err := r.triggerDelete(ctx, azVolume.Name); err != nil {
-			//If delete failed, requeue request
-			return reconcile.Result{Requeue: true}, err
+	// azVolume deletion
+	if deletionRequested(&azVolume.ObjectMeta) {
+		if azVolume.Status.State == v1alpha1.VolumeCreated || azVolume.Status.State == v1alpha1.VolumeCreationFailed || azVolume.Status.State == v1alpha1.VolumeOperationPending {
+			klog.Infof("Beginning deletion of AzVolume object")
+			if err := r.triggerDelete(ctx, azVolume.Name, azVolume); err != nil {
+				//If delete failed, requeue request
+				return reconcile.Result{Requeue: true}, err
+			}
 		}
 		//azVolume creation
 	} else if azVolume.Status.Detail == nil {
-		if azVolume.Status.Error == nil {
+		if azVolume.Status.State == v1alpha1.VolumeOperationPending || azVolume.Status.State == v1alpha1.VolumeCreationFailed {
 			klog.Infof("Creating Volume (%s)...", azVolume.Spec.UnderlyingVolume)
-			if err := r.triggerCreate(ctx, azVolume.Name); err != nil {
+			if err := r.triggerCreate(ctx, azVolume.Name, azVolume); err != nil {
```

Comment:
Instead of passing the name and object separately, why not just pass the object?